# Redo Big Run: apply small color 3-class CNN to mixed patches

This file is similar to DF_Filter.020
Use the CNN was trained on the obviously blue, green, black from DF_Filter.021      
Predict the 'mixed' patches i.e. not obviously blue, green, black from DF_Filter.021  
Generate csv file.

DF_Filter.021: 
* Read from 'raw' images. Make patches. 
* Write to 'cache/train' and 'cache/valid'.
* Write mixed uncertain images to 'patches'.
DF_Filter.022: 
* Train the CNN on {black,green,blue}={3930,2946,3410}
* use train/valid split 80:20.  
DF_Filter.023: 
* Run the CNN on 36739 files in 'patches'. 
* Write labels to 'predictions.csv' file.
* 

In [1]:
import time
import os
#  os.environ['CUDA_VISIBLE_DEVICES'] = "0"   # turns off GPU?
import glob
import cv2 # OpenCV-Python
from PIL import Image
Image.MAX_IMAGE_PIXELS = None
import numpy as np
import json
import tensorflow as tf
from tensorflow import keras
import keras.layers as kl
import pandas as pd
import matplotlib.pyplot as plt


In [2]:
DIR_IMAGES_IN = "/home/jrm/Martinez/images/patches/"
#DIR_IMAGES_IN = "/home/jrm/Martinez/images/tinytest/"   # just for testing
PATTERN_PATCHES = "*.jpg"
OUTPUT_CSV = "predictions.csv"
PATCH_SIZE=224  # matches VGG
IMAGE_SIZE=(PATCH_SIZE,PATCH_SIZE)
DIR_MODELS = "/home/jrm/Martinez/models/"
FILE_MODEL = "CNN_SMALL_REDO"
filepath=DIR_MODELS+FILE_MODEL
print("Will use model from "+filepath)
cnn = keras.models.load_model(filepath)

Will use model from /home/jrm/Martinez/models/CNN_SMALL_REDO


In [3]:
def get_image_names(path,pattern):
    paths = glob.glob(path+pattern)
    names = [os.path.basename(x) for x in paths]
    return names
FILENAMES_PATCHES = get_image_names(DIR_IMAGES_IN,PATTERN_PATCHES)

In [4]:
normalization_layer = kl.Rescaling(1.0/255.0)

In [5]:

# With keras models, augmentation and dropout are inactive during prediction.
import csv
csvpath = DIR_IMAGES_IN+OUTPUT_CSV
print("Patches to process: ",len(FILENAMES_PATCHES))
color_counts={'black':0,'blue':0,'green':0,'mixed':0}
pcount = mcount = 0
with open(csvpath, 'w', newline='') as csvfile:
    datawriter = csv.writer(csvfile, delimiter=',')
    start = time.time()
    for filename in FILENAMES_PATCHES:
        path = DIR_IMAGES_IN+filename
        img = keras.preprocessing.image.load_img(
            path, target_size=IMAGE_SIZE
        )
        img_array = keras.preprocessing.image.img_to_array(img)
        img_array = normalization_layer(img_array)
        img_array = tf.expand_dims(img_array, 0)  # create tensor
        # TODO: consider using predict_classes() instead
        predictions = cnn.predict(img_array)
        score = predictions[0]  # TODO: explore other array slices
        blackscore,bluescore,greenscore = score
        if blackscore >= 0.90:
            word = "black"
            color_counts['black'] += 1
        elif bluescore >= 0.90:
            word = "blue"
            color_counts['blue'] += 1
        elif greenscore >= 0.90:
            word = "green"
            color_counts['green'] += 1
        else:
            word = "mixed"
            color_counts['mixed'] += 1
        os.rename(path,DIR_IMAGES_IN+word+"/"+filename)
        datawriter.writerow( (filename, score[0], score[1], word) )
        pcount += 1
        mcount += 1
        if (mcount == 1000):
            print("At",pcount,color_counts)
            mcount = 0
    end = time.time()
print("Elapsed time:",end-start)
print("Total:",pcount)
print (color_counts)

Patches to process:  9567
At 1000 {'black': 396, 'blue': 401, 'green': 64, 'mixed': 139}
At 2000 {'black': 797, 'blue': 809, 'green': 120, 'mixed': 274}
At 3000 {'black': 1190, 'blue': 1229, 'green': 176, 'mixed': 405}
At 4000 {'black': 1604, 'blue': 1625, 'green': 227, 'mixed': 544}
At 5000 {'black': 2017, 'blue': 2026, 'green': 269, 'mixed': 688}
At 6000 {'black': 2457, 'blue': 2410, 'green': 307, 'mixed': 826}
At 7000 {'black': 2860, 'blue': 2816, 'green': 358, 'mixed': 966}
At 8000 {'black': 3268, 'blue': 3234, 'green': 395, 'mixed': 1103}
At 9000 {'black': 3698, 'blue': 3628, 'green': 440, 'mixed': 1234}
Elapsed time: 227.35874485969543
Total: 9567
{'black': 3932, 'blue': 3859, 'green': 472, 'mixed': 1304}


## Data subset was used to set thresholds

At .99 threshold: 78 black, 69 blue, 688 green, 650 uncertain  
At .95 threshold: 101 black, 90 blue, 864 green, 430 uncertain  
At .90 threshold: 108 black, 101 blue, 932 green, 344 uncertain  

At .9/.9/.995: 108 black, 101 blue, 585 green, 691 uncertain  
At .9/.9/.997: 108 black, 101 blue, 452 green, 824 uncertain  
At .9/.9/.998: 108 black, 101 blue, 283 green, 993 uncertain  
At .9/.9/.999: 108 black, 101 blue, 0 green, 1276 uncertain  